File: src/main/scala/uk/co/ridentbyte/view/response/ResponseOutputPane.scala
package uk.co.ridentbyte.view.response

import javafx.geometry.Insets
import javafx.scene.control.{ListView, TextArea}
import javafx.scene.layout.{GridPane, Priority}
import uk.co.ridentbyte.model.CardinalResponse
import uk.co.ridentbyte.view.util.{ColumnConstraintsBuilder, RowConstraintsBuilder}

case class ResponseOutputPane(response: CardinalResponse) extends GridPane {

  setHgap(10)
  setVgap(10)
  setPadding(new Insets(20))
  getStyleClass.addAll("plain-border", "round-border")

  getColumnConstraints.add(ColumnConstraintsBuilder().withHgrow(Priority.ALWAYS).build)
  getRowConstraints.add(RowConstraintsBuilder().withVgrow(Priority.ALWAYS).withPercentageHeight(40).build)
  getRowConstraints.add(RowConstraintsBuilder().withVgrow(Priority.ALWAYS).withPercentageHeight(60).build)

  val listHeaders = new ListView[String]()
  listHeaders.getStyleClass.add("cardinal-font")
  response.raw.headers.foreach {
    header => listHeaders.getItems.add(header._1 + ": " + header._2.mkString(""))
  }
  add(listHeaders, 0, 0)

  val textAreaBody = new TextArea()
  textAreaBody.setEditable(false)
  textAreaBody.getStyleClass.add("cardinal-font")
  textAreaBody.setText(response.formattedBody)
  add(textAreaBody, 0, 1)

}

File: src/main/scala/uk/co/ridentbyte/view/response/ResponseOutputPane.java
package uk.co.ridentbyte.view.response;

import javafx.geometry.Insets;
import javafx.scene.control.ListView;
import javafx.scene.control.TextArea;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import uk.co.ridentbyte.model.CardinalResponse;
import uk.co.ridentbyte.view.util.ColumnConstraintsBuilder;
import uk.co.ridentbyte.view.util.RowConstraintsBuilder;

import java.util.Map;

public class ResponseOutputPane extends GridPane {

    public ResponseOutputPane(CardinalResponse response) {
        this.setHgap(10);
        this.setVgap(10);
        this.setPadding(new Insets(20));
        this.getStyleClass().addAll("plain-border", "round-border");
        this.getColumnConstraints().add(new ColumnConstraintsBuilder().withHgrow(Priority.ALWAYS).build());
        this.getRowConstraints().add(new RowConstraintsBuilder().withVgrow(Priority.ALWAYS).withPercentageHeight(40).build());
        this.getRowConstraints().add(new RowConstraintsBuilder().withVgrow(Priority.ALWAYS).withPercentageHeight(60).build());

        ListView<String> listHeaders = new ListView<>();
        listHeaders.getStyleClass().add("cardinal-font");
        Map<String, String> headers = response.getHeaders();
        for (Map.Entry<String, String> header : headers.entrySet()) {
            listHeaders.getItems().add(header.getKey() + ": " + header.getValue());
        }

        // TODO - Currently manually adding Status header. Should this be somewhere else in the UI?
        listHeaders.getItems().add("Status: " + response.getStatusCode());
        listHeaders.getItems().sort((String::compareToIgnoreCase));

        this.add(listHeaders, 0, 0);

        TextArea textAreaBody = new TextArea();
        textAreaBody.setEditable(false);
        textAreaBody.getStyleClass().add("cardinal-font");
        textAreaBody.setText(response.formattedBody());
        this.add(textAreaBody, 0, 1);
    }

}
